src/read_shard.rs

```diff
@@ -1,3 +1,291 @@
-fn main() {
-    println!("Hello, world!");
+use crate::io::router::{RouterBuilder, RouterHandler};
+use anyhow::{Ok, Result};
+use messages::requests::get_client_shard_info_request::GetClientShardInfoRequest;
+use messages::requests::write_request::WriteRequest;
+use messages::responses::get_client_shard_info_response::GetClientShardInfoResponse;
+use messages::responses::write_response::WriteResponse;
+use rand::Rng;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+use tokio::time;
+mod io;
+mod messages;
+use crate::messages::{
+    requests::{
+        announce_shard_request::AnnounceShardRequest,
+        get_shared_peers_request::GetSharedPeersRequest, get_version_request::GetVersionRequest,
+        query_version_request::QueryVersionRequest, read_request::ReadRequest,
+    },
+    responses::{
+        announce_shard_response::AnnounceShardResponse,
+        get_shared_peers_response::GetSharedPeersResponse,
+        get_version_response::GetVersionResponse, query_version_response::QueryVersionResponse,
+        read_response::ReadResponse,
+    },
+};
+
+static MAIN_INSTANCE_IP_PORT: ([u8; 16], u16) =
+    ([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1], 8080);
+#[derive(Clone, Debug)]
+pub struct ReadShard {
+    _reader_ip_port: Arc<Mutex<([u8; 16], u16)>>,
+    writer_id: Arc<Mutex<u16>>,
+    writer_ip_port: Arc<Mutex<([u8; 16], u16)>>,
+    peers: Arc<Mutex<Vec<([u8; 16], u16)>>>,
+    requested_version: Arc<Mutex<u64>>,
+    current_version: Arc<Mutex<u64>>,
+    history: Arc<Mutex<Vec<(String, String)>>>,
+    data: Arc<Mutex<HashMap<String, String>>>,
+}
+
+impl RouterHandler for ReadShard {
+    fn handle_announce_shard_response(&self, res: &AnnounceShardResponse) {
+        let writer_number = res.writer_number;
+        let mut writer_id = self.writer_id.lock().unwrap();
+        *writer_id = writer_number;
+    }
+
+    fn handle_read_request(&self, req: &ReadRequest) -> ReadResponse {
+        let key = String::from_utf8_lossy(&req.key).into_owned();
+        let value = self.data.lock().unwrap().get(&key).cloned();
+        if value.is_none() {
+            return ReadResponse {
+                error: 1,
+                key: req.key.clone(),
+                value: Vec::new(),
+            };
+        } else {
+            return ReadResponse {
+                error: 0,
+                key: req.key.clone(),
+                value: value.unwrap().into_bytes(),
+            };
+        }
+    }
+
+    fn handle_get_shared_peers_response(&self, res: &GetSharedPeersResponse) {
+        let mut peers = self.peers.lock().unwrap();
+        *peers = res.peer_ips.clone();
+        let mut writer_ip_port = self.writer_ip_port.lock().unwrap();
+        *writer_ip_port = res.peer_ips[0];
```

Comment:
I don't think you actually care about keeping track of this it's not used anywhere